File: codex-rs/hepta-types/conformance/verify_rejections.py
#!/usr/bin/env python3
"""Independent raw-byte rejection oracle for Platform Types canonical V1."""
from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
DOC = json.loads(
    (ROOT / "codex-rs/hepta-types/CANONICAL_V1_CONFORMANCE.json").read_text(
        encoding="utf-8"
    )
)
DOMAIN = DOC["domain"].encode("utf-8")
MAX_BYTES = DOC["maxEncodedBytes"]
MAX_ITEMS = DOC["maxContainerItems"]
MAX_DEPTH = DOC["maxDepth"]


class Reader:
    def __init__(self, data: bytes):
        self.data = data
        self.offset = 0

    def take(self, n: int) -> bytes:
        end = self.offset + n
        if end > len(self.data):
            raise ValueError("truncated")
        out = self.data[self.offset:end]
        self.offset = end
        return out

    def n(self, size: int) -> int:
        return int.from_bytes(self.take(size), "big")

    def l16(self) -> bytes:
        return self.take(self.n(2))

    def l32(self) -> bytes:
        return self.take(self.n(4))


def ascii_text(raw: bytes, label: str) -> str:
    try:
        return raw.decode("ascii")
    except UnicodeDecodeError as exc:
        raise ValueError(label) from exc


def valid_module(raw: bytes) -> None:
    value = ascii_text(raw, "module")
    if not value:
        raise ValueError("module")
    for part in value.split("."):
        if (
            not part
            or not part[0].isalnum()
            or not part[-1].isalnum()
            or any(not (ch.islower() or ch.isdigit() or ch in "_-") for ch in part)
        ):
            raise ValueError("module")


def valid_local(raw: bytes) -> None:
    value = ascii_text(raw, "local")
    if (
        not value
        or not value[0].isalnum()
        or not value[-1].isalnum()
        or any(not (ch.islower() or ch.isdigit() or ch in "._-") for ch in value)
    ):
        raise ValueError("local")


def valid_namespaced(raw: bytes) -> None:
    if raw.count(b":") != 1:
        raise ValueError("type id")
    namespace, local = raw.split(b":", 1)
    valid_module(namespace)
    valid_local(local)


def valid_stable(raw: bytes) -> None:
    value = ascii_text(raw, "stable id")
    if not value or any(not (ch.isalnum() or ch in "._-:") for ch in value):
        raise ValueError("stable id")


def valid_label(raw: bytes) -> None:
    value = ascii_text(raw, "label")
    if (
        not value
        or len(raw) > 128
        or not value[0].isalnum()
        or not value[-1].isalnum()
        or any(not (ch.islower() or ch.isdigit() or ch in "._-:") for ch in value)
    ):
        raise ValueError("label")


def validate_value(r: Reader, depth: int = 0) -> None:
    tag = r.n(1)
    if tag == 1:
        if r.n(1) not in (0, 1):
            raise ValueError("invalid bool")
    elif tag == 2:
        r.take(8)
    elif tag == 3:
        r.take(16)
    elif tag == 4:
        r.take(8)
    elif tag == 5:
        r.l32()
    elif tag == 6:
        raw = r.l32()
        try:
            text = raw.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise ValueError("invalid text") from exc
        if "\0" in text:
            raise ValueError("NUL text")
    elif tag == 7:
        r.take(32)
    elif tag == 8:
        valid_stable(r.l16())
    elif tag in (9, 10):
        if depth >= MAX_DEPTH:
            raise ValueError("depth")
        count = r.n(4)
        if count > MAX_ITEMS:
            raise ValueError("items")
        previous = None
        for _ in range(count):
            if tag == 10:
                key = r.l16()
                valid_label(key)
                if previous is not None and key <= previous:
                    raise ValueError("map order")
                previous = key
            validate_value(r, depth + 1)
    else:
        raise ValueError("invalid tag")


def validate_raw(raw: bytes) -> None:
    if len(raw) > MAX_BYTES:
        raise ValueError("size")
    r = Reader(raw)
    if r.take(4) != b"HPTC" or r.n(2) != DOC["encodingVersion"]:
        raise ValueError("header")
    if r.l16() != DOMAIN:
        raise ValueError("domain")
    valid_namespaced(r.l16())
    if r.n(4) == 0:
        raise ValueError("schema")
    count = r.n(4)
    if count > MAX_ITEMS:
        raise ValueError("fields")
    previous = None
    for _ in range(count):
        name = r.l16()
        valid_label(name)
        if previous is not None and name <= previous:
            raise ValueError("field order")
        previous = name
        validate_value(r)
    if r.offset != len(raw):
        raise ValueError("trailing")


def u16(value: int) -> bytes:
    return value.to_bytes(2, "big")


def u32(value: int) -> bytes:
    return value.to_bytes(4, "big")


def framed16(value: bytes) -> bytes:
    return u16(len(value)) + value


def header(*, schema_version: int, field_count: int) -> bytes:
    return (
        b"HPTC"
        + u16(DOC["encodingVersion"])
        + framed16(DOMAIN)
        + framed16(b"platform.types:negative")
        + u32(schema_version)
        + u32(field_count)
    )


def bool_value(value: int) -> bytes:
    return b"\x01" + bytes([value])


def u64_value(value: int) -> bytes:
    return b"\x02" + value.to_bytes(8, "big")


def rejection_bytes(case: dict) -> bytes:
    kind = case["kind"]
    if kind == "raw_reject":
        return bytes.fromhex(case["encodingHex"])
    if kind == "duplicate_field":
        return (
            header(schema_version=case["schemaVersion"], field_count=2)
            + framed16(b"a")
            + u64_value(1)
            + framed16(b"a")
            + u64_value(2)
        )
    if kind == "duplicate_map_key":
        value = (
            b"\x0a"
            + u32(2)
            + framed16(b"a")
            + bool_value(0)
            + framed16(b"a")
            + bool_value(1)
        )
        return header(schema_version=case["schemaVersion"], field_count=1) + framed16(b"map") + value
    if kind == "zero_schema_version":
        return header(schema_version=0, field_count=0)
    if kind == "oversize_bytes":
        payload = b"\x05" + u32(case["payloadBytes"]) + bytes(case["payloadBytes"])
        return (
            header(schema_version=case["schemaVersion"], field_count=1)
            + framed16(b"payload")
            + payload
        )
    if kind == "depth_overflow":
        value = bool_value(0)
        for _ in range(case["depth"]):
            value = b"\x09" + u32(1) + value
        return (
            header(schema_version=case["schemaVersion"], field_count=1)
            + framed16(b"nested")
            + value
        )
    raise ValueError(f"unknown rejection kind: {kind}")


for case in DOC["rejections"]:
    raw = rejection_bytes(case)
    try:
        validate_raw(raw)
    except ValueError:
        continue
    raise SystemExit(f"{case['id']}: raw-byte rejection oracle accepted invalid bytes")

print(f"platform.types Python raw rejection vectors: {len(DOC['rejections'])} rejected")

File: codex-rs/hepta-types/conformance/verify_rejections.mjs
// Independent raw-byte rejection oracle for Platform Types canonical V1.
import fs from "node:fs";

const DOC = JSON.parse(
  fs.readFileSync("codex-rs/hepta-types/CANONICAL_V1_CONFORMANCE.json", "utf8"),
);
const DOMAIN = Buffer.from(DOC.domain, "utf8");
const MAX_BYTES = DOC.maxEncodedBytes;
const MAX_ITEMS = DOC.maxContainerItems;
const MAX_DEPTH = DOC.maxDepth;
const UTF8 = new TextDecoder("utf-8", { fatal: true });

class Reader {
  constructor(data) {
    this.data = data;
    this.offset = 0;
  }
  take(n) {
    const end = this.offset + n;
    if (end > this.data.length) throw new Error("truncated");
    const out = this.data.subarray(this.offset, end);
    this.offset = end;
    return out;
  }
  n(size) {
    let value = 0;
    for (const byte of this.take(size)) value = value * 256 + byte;
    return value;
  }
  l16() {
    return this.take(this.n(2));
  }
  l32() {
    return this.take(this.n(4));
  }
}

function ascii(raw, label) {
  for (const byte of raw) if (byte > 0x7f) throw new Error(label);
  return raw.toString("ascii");
}

function validModule(raw) {
  const value = ascii(raw, "module");
  const parts = value.split(".");
  if (
    parts.length === 0 ||
    parts.some((part) => !/^[a-z0-9](?:[a-z0-9_-]*[a-z0-9])?$/.test(part))
  ) {
    throw new Error("module");
  }
}

function validLocal(raw) {
  const value = ascii(raw, "local");
  if (!/^[a-z0-9](?:[a-z0-9._-]*[a-z0-9])?$/.test(value)) {
    throw new Error("local");
  }
}

function validNamespaced(raw) {
  const separator = raw.indexOf(0x3a);
  if (separator <= 0 || separator !== raw.lastIndexOf(0x3a)) {
    throw new Error("type id");
  }
  validModule(raw.subarray(0, separator));
  validLocal(raw.subarray(separator + 1));
}

function validStable(raw) {
  const value = ascii(raw, "stable id");
  if (!/^[A-Za-z0-9._:-]+$/.test(value)) throw new Error("stable id");
}

function validLabel(raw) {
  const value = ascii(raw, "label");
  if (
    raw.length === 0 ||
    raw.length > 128 ||
    !/^[a-z0-9](?:[a-z0-9._:-]*[a-z0-9])?$/.test(value)
  ) {
    throw new Error("label");
  }
}

function validateValue(r, depth = 0) {
  const tag = r.n(1);
  if (tag === 1) {
    const value = r.n(1);
    if (value !== 0 && value !== 1) throw new Error("invalid bool");
    return;
  }
  if (tag === 2) {
    r.take(8);
    return;
  }
  if (tag === 3) {
    r.take(16);
    return;
  }
  if (tag === 4) {
    r.take(8);
    return;
  }
  if (tag === 5) {
    r.l32();
    return;
  }
  if (tag === 6) {
    const raw = r.l32();
    let text;
    try {
      text = UTF8.decode(raw);
    } catch (_) {
      throw new Error("invalid text");
    }
    if (text.includes("\0")) throw new Error("NUL text");
    return;
  }
  if (tag === 7) {
    r.take(32);
    return;
  }
  if (tag === 8) {
    validStable(r.l16());
    return;
  }
  if (tag === 9 || tag === 10) {
    if (depth >= MAX_DEPTH) throw new Error("depth");
    const count = r.n(4);
    if (count > MAX_ITEMS) throw new Error("items");
    let previous = null;
    for (let i = 0; i < count; i += 1) {
      if (tag === 10) {
        const key = r.l16();
        validLabel(key);
        if (previous && Buffer.compare(key, previous) <= 0) {
          throw new Error("map order");
        }
        previous = key;
      }
      validateValue(r, depth + 1);
    }
    return;
  }
  throw new Error("invalid tag");
}

function validateRaw(raw) {
  if (raw.length > MAX_BYTES) throw new Error("size");
  const r = new Reader(raw);
  if (
    r.take(4).toString("ascii") !== "HPTC" ||
    r.n(2) !== DOC.encodingVersion
  ) {
    throw new Error("header");
  }
  if (!r.l16().equals(DOMAIN)) throw new Error("domain");
  validNamespaced(r.l16());
  if (r.n(4) === 0) throw new Error("schema");
  const count = r.n(4);
  if (count > MAX_ITEMS) throw new Error("fields");
  let previous = null;
  for (let i = 0; i < count; i += 1) {
    const name = r.l16();
    validLabel(name);
    if (previous && Buffer.compare(name, previous) <= 0) {
      throw new Error("field order");
    }
    previous = name;
    validateValue(r);
  }
  if (r.offset !== raw.length) throw new Error("trailing");
}

function u16(value) {
  const out = Buffer.alloc(2);
  out.writeUInt16BE(value);
  return out;
}

function u32(value) {
  const out = Buffer.alloc(4);
  out.writeUInt32BE(value);
  return out;
}

function framed16(value) {
  return Buffer.concat([u16(value.length), value]);
}

function header(schemaVersion, fieldCount) {
  return Buffer.concat([
    Buffer.from("HPTC", "ascii"),
    u16(DOC.encodingVersion),
    framed16(DOMAIN),
    framed16(Buffer.from("platform.types:negative", "ascii")),
    u32(schemaVersion),
    u32(fieldCount),
  ]);
}

function boolValue(value) {
  return Buffer.from([1, value]);
}

function u64Value(value) {
  const out = Buffer.alloc(9);
  out[0] = 2;
  out.writeBigUInt64BE(BigInt(value), 1);
  return out;
}

function rejectionBytes(c) {
  if (c.kind === "raw_reject") {
    return Buffer.from(c.encodingHex, "hex");
  }
  if (c.kind === "duplicate_field") {
    return Buffer.concat([
      header(c.schemaVersion, 2),
      framed16(Buffer.from("a", "ascii")),
      u64Value(1),
      framed16(Buffer.from("a", "ascii")),
      u64Value(2),
    ]);
  }
  if (c.kind === "duplicate_map_key") {
    const map = Buffer.concat([
      Buffer.from([10]),
      u32(2),
      framed16(Buffer.from("a", "ascii")),
      boolValue(0),
      framed16(Buffer.from("a", "ascii")),
      boolValue(1),
    ]);
    return Buffer.concat([
      header(c.schemaVersion, 1),
      framed16(Buffer.from("map", "ascii")),
      map,
    ]);
  }
  if (c.kind === "zero_schema_version") {
    return header(0, 0);
  }
  if (c.kind === "oversize_bytes") {
    const payload = Buffer.concat([
      Buffer.from([5]),
      u32(c.payloadBytes),
      Buffer.alloc(c.payloadBytes),
    ]);
    return Buffer.concat([
      header(c.schemaVersion, 1),
      framed16(Buffer.from("payload", "ascii")),
      payload,
    ]);
  }
  if (c.kind === "depth_overflow") {
    let value = boolValue(0);
    for (let i = 0; i < c.depth; i += 1) {
      value = Buffer.concat([Buffer.from([9]), u32(1), value]);
    }
    return Buffer.concat([
      header(c.schemaVersion, 1),
      framed16(Buffer.from("nested", "ascii")),
      value,
    ]);
  }
  throw new Error("unknown rejection kind: " + c.kind);
}

for (const c of DOC.rejections) {
  const raw = rejectionBytes(c);
  let rejected = false;
  try {
    validateRaw(raw);
  } catch (_) {
    rejected = true;
  }
  if (!rejected) {
    throw new Error(c.id + ": raw-byte rejection oracle accepted invalid bytes");
  }
}

console.log(
  "platform.types Node raw rejection vectors: " +
    DOC.rejections.length +
    " rejected",
);
